format/Message.fbs

```diff
@@ -117,6 +117,40 @@ table DictionaryBatch {
   isDelta: bool = false;
 }
 
+/// A range of field nodes, identified by their offset in the schema.
+/// The offsets are zero-indexed.
+struct FieldNodeRange {
+  /// The starting offset (inclusive)
+  start: long;
+
+  /// The ending offset (exclusive)
+  end: long;
+}
+
+/// A data header describing the shared memory layout of a "bag" of "columns".
+/// It is similar to a RecordBatch but not every top level FieldNode is required
+/// to be included in the wire payload.
+table ColumnBag {
+  /// If not provided, all field nodes are included and this payload is
+  /// identical to a RecordBatch. Otherwise the reader needs to skip
+  /// top level FieldNodes that were not included.
+  includedNodes: [FieldNodeRange];
```

Comment:
So to be clear, we can't do something like provide only a nested array - and implementations will need to validate that this only skips entire top level fields?

Reply:
Can we mark this as experimental like how Tensor does? https://github.com/apache/arrow/blob/939db7f513a56d0cab12a8479d8153e5aa2ae1df/format/Tensor.fbs#L18-L20

Reply:
In RecordBatch, the FieldNodes are listed in-order from first field node, to its children, and grandchildren, followed by the second field node. Note that if we include a top level field node we must include its children. This requirement certainly applies to array-types, and I assume it applies to nested structures -- but I have not used them enough to play with the idea.

I think there are a few options to represent which top level nodes to include.
1) encoded BitSet, but it is too easy to create degenerate cases
2) each FieldNode could include a third parameter -- but in flatbuffers this means that the struct is written down differently (I think if the struct is greater than 16B then it must be pre-written before constructing the flatbuffer table that uses it)
3) include a parallel array with field node indicating which field offset, but this would be empty for child nodes
4) what remains is  a compromise listing ranges of columns that were included -- the use case I have in mind is single-digit number of ranges almost always - but columns can be easily into the tens of dozens.

> So to be clear, we can't do something like provide only a nested array - and implementations will need to validate that this only skips entire top level fields?

I can't quite tell what solution you are proposing here. I think client implementations do end up working exactly like you are saying, though. Could you elaborate on your idea or defend an alternative approach? 

-- 

> Can we mark this as experimental like how Tensor does?

Absolutely, patch update incoming.

Reply:
Sorry, I guess what I mean is "Can we make it explicit that only top-level arrays can be skipped, and top-level arrays must be skipped as a whole", i.e. we can't "patch" a child of a nested array. (And that implementations must reject "degenerate" messages that skip, say, only one child of a struct, since that makes no sense in the first place, but should be validated.)

Reply:
Ooh. I wish there was a way to tell. I don't think you can tell that a message is a degenerate. The FieldNode themselves do not identify which node or location in the schema it belongs to. They are expected to be visited in-order (parent, children, grandchildren). The point of FieldNodeRange is that this is the list of top level field nodes only -- not including nested nodes. So, as typical, it is impossible to tell if the message was written incorrectly. You typically only find out from out-of-memory access where a node thinks it's an array but it's really just a column of uint64. 

Point is, I think I'm preventing this by not being able to describe it on the wire.

Reply:
I updated the comment. Maybe this makes the intention more obvious? Does this resolve your concern?

Reply:
Yes, thanks.

I think consumers can validate, so long as they have the schema: they would traverse the schema as usual, and look at both nodes and includedNodes. If while traversing a type they ran into a buffer not included in includedNodes, then the message is invalid.

Reply:
Ah, wait, this is a list of top level nodes/fields only. Ok, I see the comments now, thanks - that works.